packages/store/src/derived.ts

```diff
@@ -0,0 +1,115 @@
+import { Store } from './store'
+import type { Listener } from './types'
+
+interface DerivedOptions<TState> {
+  onSubscribe?: (listener: Listener, derived: Derived<TState>) => () => void
+  onUpdate?: () => void
+}
+
+export type Deps = Array<Derived<any> | Store<any>>
+
+export class Derived<TState> {
+  _store!: Store<TState>
+  rootStores = new Set<Store<unknown>>()
+  deps: Deps
+
+  // Functions representing the subscriptions. Call a function to cleanup
+  _subscriptions: Array<() => void> = []
+
+  // What store called the current update, if any
+  _whatStoreIsCurrentlyInUse: Store<unknown> | null = null
+
+  constructor(deps: Deps, fn: () => TState, options?: DerivedOptions<TState>) {
```

Comment:
would it be better if this had a flat options style like `useQuery` now does?

```ts
new Derived({
  storeKey: [],
  storeFn: () => alert("expensive")
  ...options
})
```

Reply:
Done!